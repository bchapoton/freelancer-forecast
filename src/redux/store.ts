import { configureStore } from '@reduxjs/toolkit';
import parametersReducer from './slices/ParametersSlice';
import summariesReducer from './slices/SummariesSlice';
import uiReducer from './slices/UISlice';
import { StorePreloadedState } from '../helpers/TestHelper';

export const store = storeConfigurator();

export function storeConfigurator(preloadedState?: StorePreloadedState) {
    return configureStore({
        reducer: {
            parameters: parametersReducer,
            summaries: summariesReducer,
            ui: uiReducer,
        },
        preloadedState: preloadedState,
    });
}

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
