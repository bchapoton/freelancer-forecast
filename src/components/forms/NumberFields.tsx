import { useAppDispatch } from '../../redux/hooks';
import React, {
    ChangeEvent,
    FocusEvent,
    KeyboardEvent,
    ReactNode,
    useCallback,
    useMemo,
    useRef,
    useState,
} from 'react';
import { InputAdornment, InputProps as StandardInputProps, TextField } from '@mui/material';
import { ActionCreatorWithPayload } from '@reduxjs/toolkit';
import { v4 as generateUID } from 'uuid';
import NumbersHelper from '../helpers/NumbersHelper';

type BaseNumberFieldProps = {
    dispatchAction: ActionCreatorWithPayload<number, string>;
    defaultValue?: number;
    label: string;
};

export type NumberFieldProps = BaseNumberFieldProps & {
    min?: number;
    max?: number;
    unit?: ReactNode;
};

export function NumberField({ dispatchAction, defaultValue, label, min, max, unit }: NumberFieldProps) {
    const dispatch = useAppDispatch();
    const numberDefaultValue: number | '' = defaultValue === undefined ? '' : defaultValue;
    const [internalValue, setInternalValue] = useState<number | ''>(numberDefaultValue);
    const inputRef = useRef<typeof TextField>();

    const onChangeHandler = useCallback(
        (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
            if (isNaN(parseInt(event.currentTarget.value))) {
                setInternalValue('');
            } else {
                const inputValue: number = parseInt(event.currentTarget.value);
                setInternalValue(NumbersHelper.rangedNumber(inputValue, min, max));
            }
        },
        [min, max],
    );

    const onBlurHandler = useCallback(
        (event: FocusEvent<HTMLInputElement>) => {
            const parsedValue = parseInt(event.currentTarget.value);
            if (isNaN(parsedValue)) {
                setInternalValue(numberDefaultValue);
            } else {
                dispatch(dispatchAction(parsedValue));
            }
        },
        [dispatch, dispatchAction],
    );

    const onKeyDownHandler = useCallback((event: KeyboardEvent<HTMLDivElement>) => {
        if (event.key === 'Enter') {
            event.currentTarget.blur();
        }
    }, []);

    const inputProps: Partial<StandardInputProps> | undefined = useMemo(() => {
        if (!unit) return undefined;
        return {
            endAdornment: <InputAdornment position="end">{unit}</InputAdornment>,
        };
    }, [unit]);

    return (
        <TextField
            id={generateUID()}
            label={label}
            variant="outlined"
            margin="normal"
            required
            type="number"
            fullWidth
            onChange={onChangeHandler}
            onBlur={onBlurHandler}
            onKeyDown={onKeyDownHandler}
            inputRef={inputRef}
            value={internalValue}
            InputProps={inputProps}
            size="small"
        />
    );
}

export type PercentageFieldProps = BaseNumberFieldProps;

export function PercentageField({ dispatchAction, defaultValue, label }: PercentageFieldProps) {
    return (
        <NumberField
            label={label}
            dispatchAction={dispatchAction}
            defaultValue={defaultValue}
            min={0}
            max={100}
            unit="%"
        />
    );
}

type EuroFieldProps = BaseNumberFieldProps;

export function EuroField({ dispatchAction, defaultValue, label }: EuroFieldProps) {
    return <NumberField label={label} dispatchAction={dispatchAction} defaultValue={defaultValue} min={0} unit="â‚¬" />;
}

type WeeksNumberFieldProps = BaseNumberFieldProps;

export function WeeksNumberField({ dispatchAction, defaultValue, label }: WeeksNumberFieldProps) {
    return (
        <NumberField
            label={label}
            dispatchAction={dispatchAction}
            defaultValue={defaultValue}
            min={0}
            max={52}
            unit="semaines"
        />
    );
}

type IncomingSplittingPartsFieldProps = BaseNumberFieldProps;

export function IncomingSplittingPartsField({ dispatchAction, defaultValue, label }: IncomingSplittingPartsFieldProps) {
    return (
        <NumberField label={label} dispatchAction={dispatchAction} defaultValue={defaultValue} min={1} unit="part(s)" />
    );
}
