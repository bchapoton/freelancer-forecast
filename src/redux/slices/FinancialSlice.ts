import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';
import parametersService, { TotalSummary, YearSummary } from '../../services/ParametersService';

export interface FinancialState {
    yearSummary: YearSummary;
}

const emptyTotals: TotalSummary = parametersService.buildEmptyTotalSummary();

const initialState: FinancialState = {
    yearSummary: { monthSummaries: [], totals: emptyTotals },
};

export const financialSlice = createSlice({
    name: 'financial',
    initialState,
    reducers: {
        setYearSummary: (state, action: PayloadAction<YearSummary>) => {
            state.yearSummary = action.payload;
        },
    },
});

export const { setYearSummary } = financialSlice.actions;

export const selectYearSummary: (state: RootState) => YearSummary = (state: RootState) => state.financial.yearSummary;

export default financialSlice.reducer;
