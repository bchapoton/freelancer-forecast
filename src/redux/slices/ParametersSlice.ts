import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';
import { TaxationFamilyContext } from '../../services/TaxesService';
import { DAY, DayOffContext, PartTimeContext } from '../../services/CalendarService';

export interface ParametersState {
    averageDailyRate: number;
    year: number;
    vatRate: number;
    taxAllowance: number;
    socialContribution: SocialContribution[];
    taxationFamilyContext: TaxationFamilyContext;
    saving: Saving;
    partTime: PartTimeContext;
    dayOffContext: DayOffContext;
    nonWorkingDays: DAY[];
}

const initialState: ParametersState = {
    averageDailyRate: 550,
    year: 2022,
    vatRate: 20,
    taxAllowance: 34,
    socialContribution: [
        { rate: 22.2, label: 'Prestations de services (bnc et bic) et vente de marchandises (bic)' },
        { rate: 0.2, label: 'Formation prof.liberale obligatoire' },
    ],
    taxationFamilyContext: {
        incomeSplittingParts: 3,
        taxableHouseholdRevenues: 24506,
        taxAllowance: 10,
    },
    saving: { mode: 'percentage', value: 20 },
    partTime: { percentage: 100 },
    dayOffContext: { unit: 'weeks', value: 5 },
    nonWorkingDays: [DAY.SATURDAY, DAY.SUNDAY],
};

export const parametersSlice = createSlice({
    name: 'parameters',
    initialState,
    reducers: {
        setAverageDailyRate: (state, action: PayloadAction<number>) => {
            state.averageDailyRate = action.payload;
        },
        setYear: (state, action: PayloadAction<number>) => {
            state.year = action.payload;
        },
        setVatRate: (state, action: PayloadAction<number>) => {
            state.vatRate = action.payload;
        },
        setDayOffContextValue: (state, action: PayloadAction<number>) => {
            state.dayOffContext.value = action.payload;
        },
        setPartTimeValue: (state, action: PayloadAction<number>) => {
            state.partTime.percentage = action.payload;
        },
        setTaxationFamilySplittingParts: (state, action: PayloadAction<number>) => {
            state.taxationFamilyContext.incomeSplittingParts = action.payload;
        },
        setTaxationFamilyTaxableHouseholdRevenues: (state, action: PayloadAction<number>) => {
            state.taxationFamilyContext.taxableHouseholdRevenues = action.payload;
        },
        setTaxationFamilyTaxAllowance: (state, action: PayloadAction<number>) => {
            state.taxationFamilyContext.taxAllowance = action.payload;
        },
        setSavingValue: (state, action: PayloadAction<number>) => {
            state.saving.value = action.payload;
        },
        setTaxAllowance: (state, action: PayloadAction<number>) => {
            state.taxAllowance = action.payload;
        },
        setNonWorkingDays: (state, action: PayloadAction<DAY[]>) => {
            state.nonWorkingDays = action.payload;
        },
    },
});

export const {
    setAverageDailyRate,
    setYear,
    setVatRate,
    setDayOffContextValue,
    setPartTimeValue,
    setTaxationFamilySplittingParts,
    setTaxationFamilyTaxableHouseholdRevenues,
    setTaxationFamilyTaxAllowance,
    setSavingValue,
    setTaxAllowance,
    setNonWorkingDays,
} = parametersSlice.actions;

export const selectParameters: (state: RootState) => ParametersState = (state: RootState) => state.parameters;

export default parametersSlice.reducer;

export type SocialContribution = {
    rate: number;
    label: string;
};

export type Saving = {
    mode: 'percentage';
    value: number;
};
