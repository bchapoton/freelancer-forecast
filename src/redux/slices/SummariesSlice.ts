import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';
import parametersService, { TotalSummary, YearSummary } from '../../services/ParametersService';
import { IncomeTaxesSummary } from '../../services/TaxesService';

export interface SummariesState {
    yearSummary: YearSummary;
    incomeTaxesSummary: IncomeTaxesSummary;
}

const emptyTotals: TotalSummary = parametersService.buildEmptyTotalSummary();

export const initialState: SummariesState = {
    yearSummary: { monthSummaries: [], totals: emptyTotals },
    incomeTaxesSummary: {
        taxBracketSummaries: [],
        netTaxableIncome: {
            netTaxableIncomeResult: 0,
            contractorNetTaxableIncome: 0,
            familyNetTaxableIncome: 0,
            incomeSplittingParts: 1,
        },
        total: 0,
    },
};

export const summariesSlice = createSlice({
    name: 'summaries',
    initialState,
    reducers: {
        setYearSummary: (state, action: PayloadAction<YearSummary>) => {
            state.yearSummary = action.payload;
        },
        setIncomeTaxesSummary: (state, action: PayloadAction<IncomeTaxesSummary>) => {
            state.incomeTaxesSummary = action.payload;
        },
    },
});

export const { setYearSummary, setIncomeTaxesSummary } = summariesSlice.actions;

export const selectYearSummary: (state: RootState) => YearSummary = (state: RootState) => state.summaries.yearSummary;
export const selectIncomeTaxesSummary: (state: RootState) => IncomeTaxesSummary = (state: RootState) =>
    state.summaries.incomeTaxesSummary;

export default summariesSlice.reducer;
