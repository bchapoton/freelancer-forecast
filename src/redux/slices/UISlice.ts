import { createSlice } from '@reduxjs/toolkit';
import { RootState } from '../store';

export interface UIState {
    parametersDrawerOpen: boolean;
}

const initialState: UIState = {
    parametersDrawerOpen: false,
};

export const UISlice = createSlice({
    name: 'ui',
    initialState,
    reducers: {
        toggleParametersDrawer: (state) => {
            state.parametersDrawerOpen = !state.parametersDrawerOpen;
        },
    },
});

export const { toggleParametersDrawer } = UISlice.actions;

export const isParametersDrawerOpen: (state: RootState) => boolean = (state: RootState) =>
    state.ui.parametersDrawerOpen;

export default UISlice.reducer;
