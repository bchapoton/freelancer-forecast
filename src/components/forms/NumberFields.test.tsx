import renderer, { ReactTestRenderer } from 'react-test-renderer';
import React from 'react';
import { setVatRate } from '../../redux/slices/ParametersSlice';
import { store } from '../../redux/store';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';
import { TextField } from '@mui/material';
import { NumberField } from './NumberFields';

function createRenderer(defaultValue: number): ReactTestRenderer {
    return renderer.create(
        <Provider store={store}>
            <NumberField
                label="VAT rate"
                min={min}
                max={max}
                unit="my unit"
                dispatchAction={setVatRate}
                uuid={() => '1'}
                defaultValue={defaultValue}
            />
        </Provider>,
    );
}

const min: number = 0;
const max: number = 100;

it('NumberField act correctly', () => {
    const defaultValueFromStore = store.getState().parameters.vatRate;
    const min: number = 0;
    const max: number = 100;
    const component = createRenderer(defaultValueFromStore);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // trigger some changes
    const newValues: number = 60;
    const instance = component.root;
    const found = instance.findByType(TextField);
    renderer.act(() => {
        if (found.props.onChange) {
            found.props.onChange({ target: { value: newValues } });
        }
    });
    expect(component.toJSON()).toMatchSnapshot();
    // internal value must have changed
    expect(found.props.value).toEqual(newValues);
    // the store will be updated on blur event
    expect(store.getState().parameters.vatRate).toEqual(defaultValueFromStore);

    // over max
    const overMaxValue: number = 200;
    renderer.act(() => {
        if (found.props.onChange) {
            found.props.onChange({ target: { value: overMaxValue } });
        }
    });
    expect(component.toJSON()).toMatchSnapshot();
    // internal value must be the max value
    expect(found.props.value).toEqual(max);

    // under min
    const underMinValue: number = -5;
    renderer.act(() => {
        if (found.props.onChange) {
            found.props.onChange({ target: { value: underMinValue } });
        }
    });
    expect(component.toJSON()).toMatchSnapshot();
    // internal value must be the max value
    expect(found.props.value).toEqual(min);

    // blur NumberField
    const blurValue: number = 50;
    renderer.act(() => {
        if (found.props.onBlur) {
            found.props.onBlur({ target: { value: blurValue } });
        }
    });
    expect(component.toJSON()).toMatchSnapshot();
    // check if the store changed
    expect(store.getState().parameters.vatRate).toEqual(blurValue);
});
