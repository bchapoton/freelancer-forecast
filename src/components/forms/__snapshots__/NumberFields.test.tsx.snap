// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberField act correctly 1`] = `
<div
  className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
  onKeyDown={[Function]}
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    htmlFor="1"
    id="1-label"
  >
    VAT rate
    <span
      aria-hidden={true}
      className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
    >
       
      *
    </span>
  </label>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="1"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="number"
      value={20}
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
      >
        my unit
      </p>
    </div>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-14lo706"
      >
        <span>
          VAT rate
           
          *
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`NumberField act correctly 2`] = `
<div
  className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
  onKeyDown={[Function]}
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    htmlFor="1"
    id="1-label"
  >
    VAT rate
    <span
      aria-hidden={true}
      className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
    >
       
      *
    </span>
  </label>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="1"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="number"
      value={60}
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
      >
        my unit
      </p>
    </div>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-14lo706"
      >
        <span>
          VAT rate
           
          *
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`NumberField act correctly 3`] = `
<div
  className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
  onKeyDown={[Function]}
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    htmlFor="1"
    id="1-label"
  >
    VAT rate
    <span
      aria-hidden={true}
      className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
    >
       
      *
    </span>
  </label>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="1"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="number"
      value={100}
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
      >
        my unit
      </p>
    </div>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-14lo706"
      >
        <span>
          VAT rate
           
          *
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`NumberField act correctly 4`] = `
<div
  className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
  onKeyDown={[Function]}
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    htmlFor="1"
    id="1-label"
  >
    VAT rate
    <span
      aria-hidden={true}
      className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
    >
       
      *
    </span>
  </label>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="1"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="number"
      value={0}
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
      >
        my unit
      </p>
    </div>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-14lo706"
      >
        <span>
          VAT rate
           
          *
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`NumberField act correctly 5`] = `
<div
  className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
  onKeyDown={[Function]}
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    htmlFor="1"
    id="1-label"
  >
    VAT rate
    <span
      aria-hidden={true}
      className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
    >
       
      *
    </span>
  </label>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="1"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="number"
      value={0}
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
      >
        my unit
      </p>
    </div>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-14lo706"
      >
        <span>
          VAT rate
           
          *
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
